src/App.tsx

```diff
@@ -1,24 +1,196 @@
-import React from 'react';
+import React, { useCallback, useEffect, useState } from 'react';
 import { UserWarning } from './UserWarning';
-
-const USER_ID = 0;
+import { Header } from './components/Header';
+import { Footer } from './components/Footer';
+import { TodoList } from './components/TodoList';
+import { ErrorMessage } from './components/ErrorMessage';
+import * as todosService from './api/todos';
+import { Todo } from './types/Todo';
+import { Error } from './types/Error';
+import { Status } from './types/Status';
+import { USER_ID } from './utils/USER_ID';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<Error | null>(null);
+  const [filter, setFilter] = useState<Status>(Status.All);
+  const [isDisabled, setIsDisabled] = useState(false);
+  const [temptTodo, setTemptTodo] = useState<Todo | null>(null);
+  const [query, setQuery] = useState<string>('');
+
+  const closeError = useCallback(
+    () => setError(null),
+    [],
+  );
+
+  useEffect(() => {
+    const timeoutId = setTimeout(() => {
+      setError(null);
+    }, 3000);
+
+    return () => clearTimeout(timeoutId);
+  }, [error]);
+
+  useEffect(() => {
+    todosService.getTodos()
+      .then(res => setTodos(res))
+      .catch(() => setError(Error.UnableToLoadAll));
+  }, []);
+
+  const filterTodos = useCallback((currentTodos: Todo[], status: Status) => {
+    return currentTodos.filter(todo => {
+      switch (status) {
+        case Status.All:
+          return todo;
+        case Status.Active:
+          return !todo.completed;
+        case Status.Completed:
+          return todo.completed;
+        default:
+          return todo;
+      }
+    });
+  }, []);
+
+  const filteredTodos = filterTodos(todos, filter);
+
+  const isCompleted = !!todos.filter(todo => todo.completed).length;
+
+  const handleClearCompleted = () => {
+    setTodos(todos.filter(todo => !todo.completed));
+  };
+
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const newTitle = event.target.value;
+
+    setQuery(newTitle);
+  };
+
+  const handleSubmit = (e: React.FormEvent) => {
+    e.preventDefault();
+
+    if (!query || !query.trim()) {
+      setError(Error.EmptyTitile);
+
+      return;
+    }
+
+    if (query) {
+      setIsDisabled(true);
+
+      setTemptTodo({
+        id: 0,
+        title: '',
+        completed: false,
+      });
+    }
+
+    todosService.addTodo(query.trim())
+      .then(newTodo => {
+        setTodos(currentTodos => [...currentTodos, newTodo] as Todo[]);
+        setQuery('');
+      })
+      .catch(() => {
+        setError(Error.UnableToAdd);
+      })
+      .finally(() => {
+        setTemptTodo(null);
+        setIsDisabled(false);
+      });
+  };
+
+  const deleteTodo = (id: number) => {
+    setTemptTodo({
+      ...todos.find(todo => todo.id === id),
+      id: 0,
+    } as Todo);
```

Comment:
do you need type assertion here? can it be done without it? smth like this:
```

```
```suggestion
    setTemptTodo({
      ...todos[0],
      id: 0,
    });
```